### PR TITLE
interfaces/builtin: fix custom-device default udev kernel rules

### PR DESCRIPTION
**Pull request by olivercalder**
The KERNEL value in udev rules must be the basename of the device path. For devices for which there is not a matching kernel value specified in the custom-device `udev-tagging` section, a default udev kernel rule is generated.  Previously, https://github.com/snapcore/snapd/pull/12734 (and prior) generated these default rules by using the complete device path relative to `/dev/`.  However, for device paths which are in subdirectories of `/dev/`, this means that the kernel values were not basenames, which violates the udev spec.

This commit changes this behavior to instead generate udev kernel rules using the basename of each specified device.

Since ambiguity would arise if multiple devices had the same basename, this change introduces a check to ensure that all the specified devices have unique basenames.

Additionally, this commit introduces a check to ensure that all specified kernel values in the `udev-tagging` section are basenames.

It is still the case that each specified kernel value must match one of the specified devices.

There are currently problems with `vet` where it is claimed that several of the `[]string` variables in `validateUDevDevicesUniqueBasenames()` are unused.  These variables are used in a several ways, so further investigation is required as to why this is the case.

### CODE REVIEW

**Comment by olivercalder** [comment]
I have a lingering question about the way the custom-device handles udev kernel rules (which holds for this PR, https://github.com/snapcore/snapd/pull/12734, and before). Here is a screenshot of the custom-device interface documentation:
image

The screenshot seems to suggest two things:

1. a kernel device (in this example, `full`) doesn't actually need to match one of the specified devices;
2. there shouldn't be default kernel rules generated for all specified devices based on their basenames (in this example, the longer `js...` specification)

Is it actually necessary that each `udev-tagging` kernel value matches the basename of a listed device?

Additionally, while I do still think it is necessary to generate default udev kernel rules, I'm unsure how these should be handled when regular expressions are involved. For example, in the screenshot above, the rule `KERNEL==js{[0-9],[12][0-9],3[01]}` would be automatically generated, but the rule `kernel: js[0-9]*` is specified. The specified rule is a superset of the device basename, so it should be correct from a udev point of view, but then the default and the specified rule would both refer to the same device path. Furthermore, the latter does not directly match the device path basename, so the checks for matching device path would fail and prohibit that specification. Is this the desired way for situations like this to be handled?

Note: It is not tractable to decide whether one regular expression is a subset of another in general, so that would not be a solution to the device name matching device path problem if the latter includes a regular expression -- and at the moment, no regular expression matching is done anyway, just string matching.

**Comment by olivercalder** [comment]
The unit tests are failing on this:
```
FAIL: custom_device_test.go:406: CustomDeviceInterfaceSuite.TestUDevSpec

custom_device_test.go:542:
    c.Assert(decomposedSnippets, testutil.DeepUnsortedMatches, testData.expectedRules, testLabel)
... container1 []map[string]string = []map[string]string{map[string]string{"KERNEL":"\"event[0-9]\""}, map[string]string{"KERNEL":"\"js*\""}, map[string]string{"KERNEL":"\"mice\""}, map[string]string{"KERNEL":"\"qcom,qseecom\""}}
... container2 []map[string]string = []map[string]string{map[string]string{"KERNEL":"\"event[0-9]\""}, map[string]string{"KERNEL":"\"mice\""}, map[string]string{"KERNEL":"\"js*\""}, map[string]string{"KERNEL":"\"dma_heap/qcom,qseecom\""}}
... yaml: 
... element [3]=map[KERNEL:"qcom,qseecom"] was unmatched in the second container
```
The line number stated in the failure is a comment line (the assert is 2 lines earlier), and there is no mention of `qcom` anywhere in the file, so not sure what is going on here... I'll keep investigating.

**Comment by olivercalder** [comment]
Ah, it's because https://github.com/snapcore/snapd/pull/12697 was merged in the meantime, with old assumptions about custom-device kernel rules. I'll update them to fix this issue.

**Comment by pedronis** [comment]
@olivercalder 

> a kernel device (in this example, full) doesn't actually need to match one of the specified devices;

in this scenario there won't be an apparmor rule for the device? so the device in the end will not be accessible to the snap?

File: interfaces/builtin/custom_device.go [hunk 1]
```diff
@@ -155,25 +185,26 @@ func (iface *customDeviceInterface) validateUDevTaggingRule(rule map[string]inte
...
+			// of the given devices
+			foundMatch := false
+			for _, devicePath := range devices {
+				if deviceName == filepath.Base(devicePath) {
+					foundMatch = true
+					break
 				}
 			}
+			if !foundMatch {
+				err = fmt.Errorf(`%q does not match a specified device`, deviceName)
```
Comment (pedronis) [inline interfaces/builtin/custom_device.go:206]: wouldn't s/not match a/not match any/ read slightly better?

**Review by pedronis — APPROVED** [review]
lgtm

**Comment by tsunghanliu** [comment]
I built a version of snapd with this PR for customer to verify and they said their snaps which are using devices in sub-dirs of `/dev` now work well.

**Review by mvo5 — APPROVED** [review]
Thank you for this fix!

**Pull request merged**